lens/history.go

```diff
@@ -0,0 +1,283 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore/query"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// historyItem represents an item in a particular schema's history, it
+// links to the previous and next version items if they exist.
+type historyItem struct {
+	// The schema version id of this history item.
+	schemaVersionID string
+
+	// The history item for the next schema version, if there is one
+	// (for the most recent schema version this will be None).
+	next immutable.Option[*historyItem]
+
+	// The history item for the previous schema version, if there is
+	// one (for the initial schema version this will be None).
+	previous immutable.Option[*historyItem]
+}
+
+// targetedHistoryItem represents an item in a particular schema's history, it
+// links to the previous and next version items if they exist.
+//
+// It also contains a vector which describes the distance and direction to the
+// target schema version (given as an input param on construction).
+type targetedHistoryItem struct {
+	// The schema version id of this history item.
+	schemaVersionID string
+
+	// The history item for the next schema version, if there is one
+	// (for the most recent schema version this will be None).
+	next immutable.Option[*targetedHistoryItem]
+
+	// The history item for the previous schema version, if there is
+	// one (for the initial schema version this will be None).
+	previous immutable.Option[*targetedHistoryItem]
+
+	// The distance and direction from this history item to the target.
+	//
+	// A zero value indicates that this is the target item. A positive value
+	// indicates that the target is more recent. A negative value indicates
+	// that the target predates this history item.
+	targetVector int
+}
+
+// getTargetedHistory returns the history of the schema of the given id, relative
+// to the given target schema version id.
+//
+// This includes any history items that are only known via registered
+// schema migrations.
+func getTargetedHistory(
```

Comment:
suggestion: Continuation of the naming comments.

Reply:
will do in first comment